### PR TITLE
Improve derivation of CORSIKA limits and allow to define a loss fraction per energy bin ('differential loss').

### PR DESCRIPTION
**Pull request by GernotMaier**
This pull request introduces improvements to the derivation of CORSIKA simulation limits. The main changes include support for per-axis allowed loss fractions (including differential loss per energy bin) and improved histogram plotting. The algorithm to calculate the energy threshold has been improved to a 'below-peak-fraction' approach.

A new config value allows to trigger a event loss fraction calculation per energy bin (e.g., `--differential_loss_bins_per_decade=5`). Energy bins are chosen as bins per decade, with bins aligned to the decades (e.g., lower limit to 100 GeV). The limit chosen (at this point) is the max of the differential limit.

### CODE REVIEW

File: src/simtools/production_configuration/derive_corsika_limits.py [hunk 1]
```diff
@@ -307,6 +323,81 @@ def _process_file(
...
+            continue
+        limit = _compute_limits(projected, x_bins, loss_fraction, limit_type="upper")
+        keep = np.searchsorted(x_bins, limit, side="left")
+        loss = (total - float(np.sum(projected[:keep]))) / total
+        limits.append(limit)
+        energy_centers.append(float(np.sqrt(e_low * e_high)))
+        _logger.info(
+            f"Differential {name}: E=[{e_low:.4g}, {e_high:.4g}] TeV, "
+            f"N={int(total)}, limit={limit:.4g} {unit}, loss={loss:.5f}"
+        )
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply src/simtools/production_configuration/derive_corsika_limits.py:541]: No - this is good and not too bad. It is ~20-30 lines of output per production parameter value.

File: src/simtools/sim_events/histograms.py [hunk 2]
```diff
@@ -349,14 +349,31 @@ def calculate_efficiency(trig_hist, mc_hist):
...
-        )
+
+        energy_min = self.file_info.get("energy_min", 1.0e-3 * u.TeV).to("TeV").value
+        energy_max = self.file_info.get("energy_max", 1.0e3 * u.TeV).to("TeV").value
+        energy_min = max(energy_min, 1e-3)
+        energy_max = max(energy_max, 10 * energy_min)
+
+        lower_decade = np.floor(np.log10(energy_min))
+        upper_decade = np.ceil(np.log10(energy_max))
+        if upper_decade <= lower_decade:
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply src/simtools/sim_events/histograms.py:391]: Fixed - actually a very good point!

File: src/simtools/sim_events/histograms.py [hunk 3]
```diff
@@ -349,23 +362,46 @@ def calculate_efficiency(trig_hist, mc_hist):
...
+        Add 50 m for improved plotting.
+        """
         if "core_distance_bin_edges" in self.histograms:
             return self.histograms["core_distance_bin_edges"]
+
         return np.linspace(
             self.file_info.get("core_scatter_min", 0.0 * u.m).to("m").value,
-            self.file_info.get("core_scatter_max", 1.0e5 * u.m).to("m").value,
+            self.file_info.get("core_scatter_max", 1.0e5 * u.m).to("m").value + 50.0,
             100,
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply src/simtools/sim_events/histograms.py:410]: Fixed.

File: src/simtools/visualization/plot_simtel_event_histograms.py [hunk 4]
```diff
@@ -306,15 +251,21 @@ def _plot_data(ax, data, bins, plot_type, plot_params, colorbar_label):
...
 
+    for x_value in np.atleast_1d(lines.get("x", [])):
+        ax.axvline(x_value, color="r", linestyle="--", linewidth=0.5)
+
+    for y_value in np.atleast_1d(lines.get("y", [])):
+        ax.axhline(y_value, color="r", linestyle="--", linewidth=0.5)
+
+    curve = lines.get("curve")
+    if curve and curve.get("x") and curve.get("y"):
+        ax.plot(curve["x"], curve["y"], color="tab:orange", linestyle="-", linewidth=1.0)
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply src/simtools/visualization/plot_simtel_event_histograms.py:267]: This shouldn't be an issue.

**Comment by GernotMaier** [comment]
Attached the output from several zenith angles / array layouts as obtained for prod6 North. I will add a small follow up PR  to plot the limits vs zenith from the ecsv file.

[North-corsika-limits.tar.gz](https://github.com/user-attachments/files/28102861/North-corsika-limits.tar.gz)

File: src/simtools/production_configuration/derive_corsika_limits.py [hunk 5]
```diff
@@ -487,73 +799,6 @@ def _is_close(value, reference, warning_text):
...
-    return _is_close(
-        viewcone_limit,
-        histograms.file_info["viewcone_max"].to("deg")
-        if "viewcone_max" in histograms.file_info
-        else None,
-        "Upper viewcone limit is equal to the maximum viewcone distance of",
-    )
+def _core_distance_ground_to_shower(core_distance, _zenith):
+    """Return unchanged core distance; limits are handled in shower coordinates."""
+    return core_distance
```
Comment (EshitaJoshi) [inline src/simtools/production_configuration/derive_corsika_limits.py]: This function is doing nothing

File: src/simtools/production_configuration/derive_corsika_limits.py [hunk 6]
```diff
@@ -433,40 +692,93 @@ def _compute_limits(hist, bin_edges, loss_fraction, limit_type="lower"):
...
+    threshold_fraction : float, optional
+        Fraction of the stable peak used as threshold.
+
+    Returns
+    -------
+    float
+        Derived energy threshold.
+    """
+    counts = np.asarray(counts, dtype=float)
+    bin_edges = np.asarray(bin_edges, dtype=float)
```
Comment (EshitaJoshi) [inline src/simtools/production_configuration/derive_corsika_limits.py:738]: Might also be worth doing an ```np.isfinite``` check to make sure that counts and bin_edges dont have nans or infs.
Also what happens now if all counts passed are zero? That should probably raise an error somewhere

File: src/simtools/production_configuration/derive_corsika_limits.py [hunk 7]
```diff
@@ -171,6 +191,72 @@ def _resolve_telescope_configs(args_dict):
...
+                parsed[axis_name] = {
+                    "loss_fraction": fraction,
+                    "loss_min_events": min_events,
+                }
+            continue
+
+        if axis_name not in LOSS_AXES:
+            raise ValueError(
+                "Invalid axis for --allowed_losses. Allowed axes: "
+                "core_distance, angular_distance, all."
```
Comment (EshitaJoshi) [inline src/simtools/production_configuration/derive_corsika_limits.py]: Very minor, but probably better to use ```LOSS_AXES``` here instead of hard-coding the axes names in the error

File: src/simtools/production_configuration/derive_corsika_limits.py [hunk 8]
```diff
@@ -171,6 +191,72 @@ def _resolve_telescope_configs(args_dict):
...
+        if len(parts) != 3:
+            raise ValueError(
+                f"Invalid --allowed_losses value '{raw_value}'. "
+                "Expected format: axis,fraction,min_events"
+            )
+
+        axis_raw, fraction_raw, min_events_raw = parts
+        try:
+            fraction = float(fraction_raw)
+            min_events = int(min_events_raw)
```
Comment (EshitaJoshi) [inline src/simtools/production_configuration/derive_corsika_limits.py:225]: Should add a validation check somewhere that ```0 < fraction < 1``` and ```min_events > 0```

File: src/simtools/production_configuration/derive_corsika_limits.py [hunk 9]
```diff
@@ -307,7 +420,129 @@ def _process_file(
...
+            "unit": "deg",
+        },
+    }
+
+    per_axis_limits = {}
+
+    for axis_name, config in axis_configs.items():
+        if bins_per_decade > 0:
+            low = int(np.floor(np.log10(np.min(histograms.energy_bins))))
+            high = int(np.ceil(np.log10(np.max(histograms.energy_bins))))
```
Comment (EshitaJoshi) [inline src/simtools/production_configuration/derive_corsika_limits.py]: This can be moved outside the for loop, right?

**Review by EshitaJoshi — APPROVED** [review]
Minor comments only, thanks for the PR!

File: src/simtools/production_configuration/derive_corsika_limits.py (hunk 5 above)
Comment (GernotMaier) [reply src/simtools/production_configuration/derive_corsika_limits.py]: Yes - dead code accidentally left. Removed.

File: src/simtools/production_configuration/derive_corsika_limits.py (hunk 6 above)
Comment (GernotMaier) [reply src/simtools/production_configuration/derive_corsika_limits.py:738]: I don't think we need to check here the bin edges. They are defined in EventDataHistograms::energy_bins and we would obtain errors for !np.isifinite bin edges.

Added a check for zero counts.

File: src/simtools/production_configuration/derive_corsika_limits.py (hunk 7 above)
Comment (GernotMaier) [reply src/simtools/production_configuration/derive_corsika_limits.py]: Good point, fixed.

File: src/simtools/production_configuration/derive_corsika_limits.py (hunk 9 above)
Comment (GernotMaier) [reply src/simtools/production_configuration/derive_corsika_limits.py]: Yes, fixed.
